### PR TITLE
bpo-46543: add sys._getcaller

### PR DESCRIPTION
**Pull request by JelleZijlstra**
https://bugs.python.org/issue46543

### CODE REVIEW

File: Doc/library/sys.rst [hunk 1]
```diff
@@ -758,6 +758,26 @@ always available.
       It is not guaranteed to exist in all implementations of Python.
 
 
+.. function:: _getfunc(depth=1)
```
Comment (markshannon) [inline Doc/library/sys.rst]: Should default to `0` to match `sys._getframe()`.

File: Python/sysmodule.c [hunk 2]
```diff
@@ -1799,6 +1799,49 @@ sys_getallocatedblocks_impl(PyObject *module)
...
+    /
+
+Return a function object from the call stack.
+
+If optional integer depth is 1 or more, return the function object that many
+calls below the top of the stack.  If that is deeper than the call
+stack, ValueError is raised.  If depth is 0, return the current function
+object.
+
+This is similar to sys._getframe() but cheaper because it does not
```
Comment (markshannon) [inline Python/sysmodule.c]: Probably better to explain `sys._getframe()` in terms of `sys._getfunc`. `sys._getfunc` (or whatever it ends up being called) should be more portable.

**Review by markshannon — COMMENTED** [review]
Implementation looks sound. A couple of minor issues.

**Comment by markshannon** [comment]
I don't like the name `_getfunc`, as explained in https://github.com/faster-cpython/ideas/discussions/238#discussioncomment-2125650

**Comment by JelleZijlstra** [comment]
Thanks, I'll go back to `_getcaller` based on the discussion on faster-cpython.

File: Doc/library/sys.rst (hunk 1 above)
Comment (JelleZijlstra) [reply Doc/library/sys.rst]: Done

File: Python/sysmodule.c (hunk 2 above)
Comment (JelleZijlstra) [reply Python/sysmodule.c]: Added a note to the docstring and RST docs for `_getframe` instead.

**Comment by JelleZijlstra** [comment]
@markshannon would you still like this to make it into 3.11?

**Comment by Fidget-Spinner** [comment]
@JelleZijlstra I don't know about Mark, but personally I'd like to see it in 3.11. Would my review be enough? (I've held off reviewing since it's draft).

**Comment by markshannon** [comment]
Unfortunately there are a few problems with this, which stem from `function.__code__` being mutable.

1. Promoting the use of `sys._getcaller` instead of `sys._getframe` is flawed when you want the code object, as the caller function's `__code__` attribute may have changed so that `sys._getcaller(1).__code__ != sys.getframe(1).f_code` which would be surprising. Changing the `__code__` attribute of a function is a weird thing to do, so this might be OK.
2. As internal structs cease to be part of the C-API, we may be able to implement https://github.com/faster-cpython/ideas/issues/111 which would mean that a reference to the caller function would no longer be available.
3. In terms of Python semantics, a frame is defined by the (code, globals, builtins, locals) quad, not by the function from which it originally derived. The API should probably reflect that.


This was my idea originally, sorry for not spotting these flaws earlier.

**Pull request closed without merging**